### PR TITLE
Improved LED error states

### PR DESCRIPTION
**Pull request by alastairong**
Implements more LED colors for diagnosing HoloPort error states

### CODE REVIEW

**Comment by alastairong** [comment]
Tested. HPOS has non-zero exit code with following logs:

```
-- Logs begin at Wed 2020-07-22 14:56:10 UTC. --
Jul 22 14:56:15 hpos systemd[1]: hpos-led-manager.service: Main process exited, code=exited, status=1/FAILURE
Jul 22 14:56:16 hpos aorura-cli[749]: flash:blue
Jul 22 14:56:16 hpos systemd[1]: hpos-led-manager.service: Failed with result 'exit-code'.
Jul 22 14:56:16 hpos systemd[1]: hpos-led-manager.service: Consumed 29ms CPU time, received 0B IP traffic, sent 324B IP traffic.
Jul 22 15:01:00 hpos systemd[1]: Started hpos-led-manager.service.
Jul 22 15:01:00 hpos hpos-led-manager[1573]: Error: No such file or directory (os error 2)
Jul 22 15:01:00 hpos systemd[1]: hpos-led-manager.service: Main process exited, code=exited, status=1/FAILURE
Jul 22 15:01:01 hpos aorura-cli[1601]: flash:blue
Jul 22 15:01:01 hpos systemd[1]: hpos-led-manager.service: Failed with result 'exit-code'.
Jul 22 15:01:01 hpos systemd[1]: hpos-led-manager.service: Consumed 8ms CPU time, received 60B IP traffic, sent 164B IP traffic.

```

**Comment by alastairong** [comment]
Need to rethink how this works / improve error handling - currently on a hard reset / pre-registration hpos-led-manager fails and exits because it cannot find the .nix-revision file (since that was deleted)

File: overlays/holo-nixpkgs/hpos-led-manager/src/main.rs [hunk 1]
```diff
@@ -29,18 +32,48 @@ struct Args {
     flag_state: PathBuf,
 }
 
-fn main() -> Fallible<()> {
+#[tokio::main]
+async fn get_hydra_revision() -> anyhow::Result<serde_json::Value> {
```
Comment (filalex77) [inline overlays/holo-nixpkgs/hpos-led-manager/src/main.rs:36]: I think it's more logical to annotate `fn main` with `#[tokio::main]`.
```suggestion
async fn get_hydra_revision() -> anyhow::Result<serde_json::Value> {
```

File: overlays/holo-nixpkgs/hpos-led-manager/src/main.rs [hunk 2]
```diff
@@ -29,18 +32,48 @@ struct Args {
...
+        .header(header::ACCEPT, "application/json")
+        .header(header::CONTENT_TYPE, "application/json")
+        .send()
+        .await?;
+
+    let json = res.json().await?;
+    Ok(json)
+}
+
+fn main() -> Result<()> {
```
Comment (filalex77) [inline overlays/holo-nixpkgs/hpos-led-manager/src/main.rs:60]:
```suggestion
#[tokio::main]
async fn main() -> Result<()> {
```

File: overlays/holo-nixpkgs/hpos-led-manager/src/main.rs [hunk 3]
```diff
@@ -53,17 +86,53 @@ fn main() -> Fallible<()> {
 
         let hpos_config_found = Path::new("/run/hpos-init/hpos-config.json").exists();
 
-        let state = match (online, hpos_config_found) {
-            (false, _) => State::Flash(Color::Purple),
-            (true, false) => State::Static(Color::Blue),
+        if counter % HYDRA_POLLING_INTERVAL == 0 {  
+            revision_json = get_hydra_revision()?;
```
Comment (filalex77) [inline overlays/holo-nixpkgs/hpos-led-manager/src/main.rs:90]: This means we should `await` this now.

```suggestion
            revision_json = get_hydra_revision().await?;
```

File: overlays/holo-nixpkgs/hpos-led-manager/src/main.rs [hunk 4]
```diff
@@ -53,17 +86,53 @@ fn main() -> Fallible<()> {
...
+        let update_required = local_revision != hydra_revision; // If this lights up then it's likely the updater isn't working properly
+
+        let mut tls_certificate_valid = false;
+        let tls_certificate_found = Path::new("/var/lib/acme/default/account_reg.json").exists();
+        if tls_certificate_found {
+            let tls_certificate_raw = fs::read_to_string("/var/lib/acme/default/account_reg.json")
+            .expect("Something went wrong reading certificate");
+            let tls_certificate: Value = serde_json::from_str(&tls_certificate_raw)?;
+            tls_certificate_valid = tls_certificate["body"]["status"] == "valid"; // Note that the TLS_certificate returns a borrow (&value)
+        }
```
Comment (filalex77) [inline overlays/holo-nixpkgs/hpos-led-manager/src/main.rs:109]: Avoid using `mut`able variables if you can.
Remember that `if` is also an expression, and has a value.

```suggestion
        let tls_certificate_found = Path::new("/var/lib/acme/default/account_reg.json").exists();
        let tls_certificate_valid = if tls_certificate_found {
            let tls_certificate_raw = fs::read_to_string("/var/lib/acme/default/account_reg.json")
            .expect("Something went wrong reading certificate");
            let tls_certificate: Value = serde_json::from_str(&tls_certificate_raw)?;
            tls_certificate["body"]["status"] == "valid" // Note that the TLS_certificate returns a borrow (&value)
        } else {
        	false
        }; 
```

File: overlays/holo-nixpkgs/hpos-led-manager/src/main.rs [hunk 5]
```diff
@@ -53,17 +86,53 @@ fn main() -> Fallible<()> {
...
+            update_required,
+            !hpos_config_found,
+            !tls_certificate_valid,
+        ) {
+            (true, _, _, _, _, _) => State::Flash(Color::Red),
+            (false, true, _, _, _, _) => State::Flash(Color::Yellow),
+            (false, false, true, _, _, _) => State::Flash(Color::Purple),
+            (false, false, false, true, _, _) => State::Static(Color::Blue),
+            (false, false, false, false, true, _) => State::Flash(Color::Green),
+            (false, false, false, false, false, true) => State::Static(Color::Green),
```
Comment (filalex77) [inline overlays/holo-nixpkgs/hpos-led-manager/src/main.rs:127]: I find this table a bit hard to read, but I don't see any better way for now.

File: overlays/holo-nixpkgs/hpos-led-manager/src/main.rs [hunk 6]
```diff
@@ -53,17 +86,53 @@ fn main() -> Fallible<()> {
...
             _ => State::Aurora,
         };
 
         if state != state_prev {
-            led.set(state)?;
+            led.set(state).map_err(|e| e.compat())?;
             state_prev = state;
         }
-
+
```
Comment (filalex77) [inline overlays/holo-nixpkgs/hpos-led-manager/src/main.rs:135]: Trailing whitespace is evil.

```suggestion

```

**Comment by alastairong** [comment]
This adds a lot of bloat to the LED manager and is probably not worth it

**Pull request closed without merging**